SirsiDynix: EBSCO kb integration

queryluke opened the pull request:
This PR allows users to search EBSCO Knowledge base and import titles & package information from it.

## Note
I've added the necessary upgrade code to this branch. If I shouldn't have I can remove it and put the necessary sql update files in a `[module]/install/protected/2.1.0/` dir, as proposed in #318 & #319.

## Testing

### Installing 
#### Fresh Install
If installing a fresh instance of Coral, using this branch, you don't need to do anything extra. It will add the necessary columns during install of 2.1.0

#### Upgrade
If testing on a current coral install, checkout this branch. In the browser, navigate to [coral.url]/install/ to kick off the upgrade process

#### Manual
If you'd like to manually run the db changes, you can run the following
`$ mysql -u$DB_USERNAME -p$DB_PASSWORD coral_resources < $CORAL_INSTALL_DIR/resources/install/protected/update_2.1.0.sql`
and
`$ mysql -u$DB_USERNAME -p$DB_PASSWORD coral_resources < $CORAL_INSTALL_DIR/organizations/install/protected/update_2.1.0.sql`

### Setting up EBSCO Kb Config
1. You'll need to have an EBSCO KB api account set up. Or you can contact myself or Carla Clark and we should be able to pass along Sirsi's config
2. Click "Admin" > "EBSCO Kb Configuration"
3. Click the Enable checkbox and add your Customer ID and API key.
![screen shot 2017-12-12 at 7](https://user-images.githubusercontent.com/12414786/33916044-d47b721e-df74-11e7-888c-fa78764738d8.png)
4. Click Save
5. The EBSCO Kb Navigation link should appear

### Searching and browsing
1. Click on the EBSCO Kb nav link
2.  Try searches for Vendors, Packages, and Titles using the various filters.

#### Vendor search
You can click view packages to view all the packages for that vendor
![08-vendor-search](https://user-images.githubusercontent.com/12414786/33916305-1885330e-df76-11e7-8c9f-ea7a884ddbb3.png)

#### Package search
You can import (further down), or view all titles within a package
![09-package-search](https://user-images.githubusercontent.com/12414786/33916324-2fc29aac-df76-11e7-8021-3d304c4291ea.png)

#### Title search
You can import a single title, or click on the title to see the title details.
![01-title-search](https://user-images.githubusercontent.com/12414786/33916347-4866e0fe-df76-11e7-80b4-bbf0d650c251.png)
![04-title-details](https://user-images.githubusercontent.com/12414786/33916350-4e22e3da-df76-11e7-99c3-f339b3dbacf3.png)

### Importing Titles
When importing a single title, you can add select the provider, resource format, acquisition type, and add notes. To test, import a few titles and make sure those values are set.

Additionally, make sure the following values are automatically imported (which you can view from the title details):
- Title
- Description
- Any ISXNs
- Subjects
- Resource Type
- Urls (if the title has more than 2 urls available, the remaining urls will be added into the access notes)
- Coverage Statement from all available packages
<img width="454" alt="screen shot 2017-12-12 at 8 00 45 pm" src="https://user-images.githubusercontent.com/12414786/33916553-69c0f086-df77-11e7-9cc5-35618a039618.png">
<img width="460" alt="screen shot 2017-12-12 at 8 01 57 pm" src="https://user-images.githubusercontent.com/12414786/33916558-7025ea94-df77-11e7-83be-d04a8d7e33af.png">

### Importing Packages
Packages have more importing options:
- **Provider** allows you to import the provider as either a new provider, and alias of an existing provider, a parent or child of an existing provider, or not to import the provider and use an existing provider instead.
- **Package Titles** allows you to choose between import only selected titles, all titles, or no titles from the package
- **Workflow Options** allows you to start a workflow for only the package or the package and all the titles
- Format Type & Acquisition Type, these options will be applied to the package and all the titles
- Package Resource Type, what resource type you apply to the package (title resource types will be automatically imported)
- Notes
<img width="790" alt="screen shot 2017-12-12 at 8 04 21 pm" src="https://user-images.githubusercontent.com/12414786/33917358-d2ce4926-df7b-11e7-8ab6-c538ec9379a6.png">

Once all options are set, import the package. Note that it takes about 1 minute per 100 titles, even with the batching.
![11-batch-importing](https://user-images.githubusercontent.com/12414786/33917363-dad14f9c-df7b-11e7-9650-583a706d983c.png)

Once complete, check that the values were imported correctly the titles. Also check:
1. The titles have the package as a parent
2. The provider (Organization) was correctly imported, aliases, or had a hierarchy added (based on your choice)

Finally, go back to the EBSCO Kb search and browse. Now the titles/packages you've imported should have check marks indicating they've been imported.
![01-title-search](https://user-images.githubusercontent.com/12414786/33917430-4b1b496a-df7c-11e7-85a8-d66f7bb5b3af.png)

If there are any bugs or minor improvements that need to be made, please let me know.

fondrenlibrary commented:
According to the discussions on issue #302, it's preferred to rename update_2.1.0.sql to 001-334.sql (sequence no-PRno.sql)

t4k commented:
@queryluke: Could you resolve the merge conflict that has arisen?

queryluke commented:
sure thing

queryluke commented:
@t4k should be good now. Keep me posted if there are issues with any of my other PRs

jeffnm commented:
I had a successful installation (upgrade path). I've passed it over to the eresource staff for testing the actual features and will report back with any bugs or issues discovered.

fondrenlibrary commented:
Installed and quickly tested, it looks working as expected. ISBN/ISSN is supposed not imported?

queryluke commented:
@fondrenlibrary ISxNs should be imported. Are they not? Seem to be working on my local install. But I can try a fresh install and double check? Did you have a specific title?

fondrenlibrary commented:
Sometimes it works. If the EBSCO KB Title Details popup can capture and show ISXNs, they will be loaded, but sometimes the BSCO KB Title Details popup shows not ISXNs at all.
![image](https://user-images.githubusercontent.com/11461564/34686154-5c84a2e2-f470-11e7-838c-9ab909d6d91d.png)

Here is the one - Chinese Entrepreneurship in a Global Era (Chinese worlds)

queryluke commented:
Thanks for the clarification. Some of the titles in EBSOC Kb lack ISxN information. In instances where the titles don't have any ISxN information, nothing will show up in the Title Details popup and nothing will get imported. In essence, EBSCO Kb has incomplete data on the title.

fondrenlibrary commented:
Sorry, like this one,Information Systems Development: Reflections, Challenges and New Directions, it has two standard numbers from EBSCO, but the ISXN numbers can't be imported and the Title Details popup sees empty under the ISXN link.

caltechheather commented:
We've added the module and worked with EBSCO to get what should be the proper API key (something called an RM-API.) We are still working with them to verify no change in our customer ID, but there is no indication if our API key is valid or not. A search shows "Processing" with no timeout or other error messages.

techsvcslib commented:
Heather,

Does getting this API require anything special, i.e. does it require you to be an EDS subscriber? Or can anyone who’s an EBSCO customer, e.g. for EBSCOhost databases, get it?

Steve



From: caltechheather <notifications@github.com>
Reply-To: coral-erm/coral <reply@reply.github.com>
Date: Monday, January 8, 2018 at 2:27 PM
To: coral-erm/coral <coral@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [coral-erm/coral] SirsiDynix: EBSCO kb integration (#334)


We've added the module and worked with EBSCO to get what should be the proper API key (something called an RM-API.) We are still working with them to verify no change in our customer ID, but there is no indication if our API key is valid or not. A search shows "Processing" with no timeout or other error messages.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/coral-erm/coral/pull/334#issuecomment-356085958>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AC9mt4VJbqDQmgwa4gkJFm5-_VsKhy51ks5tInoogaJpZM4Q_PnS>.

queryluke commented:
@fondrenlibrary I see what you mean now. The search results show the ISxNs, but the pop-up does not. I'll take a look at what's going on.

queryluke commented:
@caltechheather If the customerID or API key are not valid, you should get a "No Results Found". But the server should also throw an Exception that explains why the search failed. Unfortunately, I think the only thing that EBSCO returns in "Authentication failed", but this should at least show up in your logs.

If you're seeing a constant spinning, it's might be because your php server has the `displayErrors` configuration set to on. So when the search fails, the server returns a 500 error which is causing the javascript to balk.

queryluke commented:
@caltechheather It looks like anyone with a customer ID should be able to get an API key:
https://developer.ebsco.com/requestapikey

techsvcslib commented:
Thanks, this is just what I needed to know.

Steve


From: Luke Aeschleman <notifications@github.com>
Reply-To: coral-erm/coral <reply@reply.github.com>
Date: Monday, January 8, 2018 at 3:28 PM
To: coral-erm/coral <coral@noreply.github.com>
Cc: Steve Oberg <steve.oberg@wheaton.edu>, Comment <comment@noreply.github.com>
Subject: Re: [coral-erm/coral] SirsiDynix: EBSCO kb integration (#334)


@caltechheather<https://github.com/caltechheather> It looks like anyone with a customer ID should be able to get an API key:
https://developer.ebsco.com/requestapikey

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/coral-erm/coral/pull/334#issuecomment-356102046>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AC9mt0hNvtkgX-5veto9vo_lcdCq25aTks5tIohfgaJpZM4Q_PnS>.

t4k commented:
@queryluke I figured out the problem on the server that @caltechheather is working on.

I checked the logs and we were getting the error:
> PHP Fatal error:  Uncaught Error: Call to undefined function curl_init() in …/resources/admin/classes/domain/EbscoKbService.php:275 …

I installed `php-curl` with `sudo apt-get install php-curl` and it works.

queryluke commented:
@fondrenlibrary Unfortunately, this seems to be an issue with EBSCO Kb.

According to the API documentation, the Title Objects should have an identifiersList which is an array of [identifiers](https://developer.ebsco.com/reference/rmref#identifier)

If you look at the `type` attribute, you'll see that values `0` & `1` correspond to ISSN & ISBN, respectively. Within the code, I [filter the identifier lists on these values](https://github.com/queryluke/coral/blob/ebscokb-integration/resources/templates/ebscoKbTitleList.php#L36).

But for whatever reason, the Title Object information I'm getting from EBSCO Kb is not the same:

__In the title list__:
```
Array
(
    [0] => Array
        (
            [id] => 978-1-4614-4950-8
            [source] => ResourceIdentifier
            [subtype] => 1
            [type] => 1
        )

    [1] => Array
        (
            [id] => 978-1-4614-4951-5
            [source] => ResourceIdentifier
            [subtype] => 2
            [type] => 1
        )

    [2] => Array
        (
            [id] => 655687
            [source] => ResourceIdentifier
            [subtype] => 0
            [type] => 7
        )

    [3] => Array
        (
            [id] => GWFE
            [source] => MFS
            [subtype] => 0
            [type] => 8
        )

    [4] => Array
        (
            [id] => 2135193
            [source] => AtoZ
            [subtype] => 0
            [type] => 9
        )
)
```
__In the title popup__:
```
Array
(
    [0] => Array
        (
            [id] => 2135193
            [source] => AtoZ
            [subtype] => 0
            [type] => 9
        )

    [1] => Array
        (
            [id] => 655687
            [source] => ResourceIdentifier
            [subtype] => 0
            [type] => 16
        )

    [2] => Array
        (
            [id] => 978-1-4614-4950-8
            [source] => ResourceIdentifier
            [subtype] => 1
            [type] => 6
        )

    [3] => Array
        (
            [id] => 978-1-4614-4951-5
            [source] => ResourceIdentifier
            [subtype] => 2
            [type] => 8
        )

    [4] => Array
        (
            [id] => GWFE
            [source] => MFS
            [subtype] => 0
            [type] => 17
        )
)
```

The alternative would be to try and parse the identifiers with some regexp or other pattern matching, but I feel this should be reported to EBSCO first to see if they can fix it on their end. I'll put in a ticket with them.

queryluke commented:
@t4k great!

t4k commented:
In our test instance we don't have the recommended database names in place. The hardcoded `coral_organizations` database in `resources/ajax_forms/getEbscoKbPackageImportForm.php` is causing a PHP fatal error because of this. (Our database name in this case is `coraltest_organizations`.)

Is there a way to get the database name from the configuration? Tying the functionality to the expected table names is probably not a good idea, especially because in previous versions of CORAL the recommended names were like `coral_organizations_prod` which would break this as well. Many people may have upgraded without changing database names.

queryluke commented:
@t4k Sure thing. I'll look into it tomorrow.

queryluke commented:
@t4k Ok, that's fixed now. Was an oversight on my part, the import scripts were using the DB name from the config. Not sure why I didn't apply that same method here.

t4k commented on resources/ajax_htmldata/getAdminEbscoKbConfigDisplay.php:
```diff
@@ -0,0 +1,21 @@
+    <div style="width: 450px; margin: 10px 0;">
+        <label for="ebscoKbEnabled">
+            <input type="checkbox" name="enabled" id="ebscoKbEnabled" value="true" <?php echo $config->settings->ebscoKbEnabled == 'Y' ? 'checked' : ''; ?>>
+            <?php echo _("Enable EBSCO Knowkledge Base"); ?>
```
typo: "Knowkledge"

queryluke replied:
fixed

t4k commented on resources/ebsco_kb.php:
```diff
@@ -0,0 +1,232 @@
+//except we don't want it to retain if they press the 'index' button
+//check what referring script is
+
+CoralSession::set('ref_script', $currentPage);
```
I got the message
> PHP Notice:  Undefined variable: currentPage

when loading the ebsco_kb.php page.

queryluke replied:
fixed

t4k commented on resources/ebsco_kb.php:
```diff
@@ -0,0 +1,232 @@
+                                        <strong><?php echo _("Name (contains)");?></strong>
+                                    </label>
+                                    <br />
+                                    <?php echo Html::text_search_field_tag('search', isset($search['name']) ? $search['name'] : '' ); ?>
```
I got the message
> PHP Deprecated:  Non-static method Html::text_search_field_tag() should not be called statically

when loading the ebsco_kb.php page.

queryluke replied:
fixed. But FYI - I was following the convention on the [index page](https://github.com/coral-erm/coral/blob/development/resources/index.php#L77), which uses this a lot.

t4k commented on resources/templates/ebscoKbTitleList.php:
```diff
@@ -0,0 +1,63 @@
+            <td>
+                <ul style="list-style: none; font-size: .8em">
+                    <?php
+                    foreach($item->identifiersList as $identifier){
```
I get the message
> PHP Warning:  Invalid argument supplied for foreach()

repeatedly when doing a search.

queryluke replied:
I'm not seeing this error. Is there a particular search that's causing it or any search?

Also what version of php are you using? Just so I can make my dev environment the same as yours. Worse case scenario, I can throw a guard around the foreach loop, but I wanted to make sure it's not an issue with the model's getter functions.

t4k replied:
Environment:
>Server version: Apache/2.4.18 (Ubuntu)
mysql Ver 14.14 Distrib 5.7.18, for Linux (x86_64) using EditLine wrapper
PHP 7.1.6-1~ubuntu16.04.1+deb.sury.org+1 (cli) (built: Jun 9 2017 08:26:34) ( NTS )

I did the same "nursing" title search as you, but we return many books which seem to be the problem. See screenshot: https://screenshots.firefox.com/TSdNqbqc0oTO5UuS/coral.vm

…And I was just informed that these come from a custom package of ours that has potentially unpredictable data in some fields. We are turning that off in our EDS, and I will do the search again and report back if it all goes away, but maybe this is a potential problem with custom packages that needs to be known about.

queryluke replied:
I see. Let me check the API and see if there is a way to distinguish between custom and managed packages. Otherwise, I can add some guards on those attributes to return empty arrays

queryluke replied:
just pushed a fix. Let me know if it's not working.

t4k commented:
Regarding `Html::text_search_field_tag`…
> I was following the convention on the index page, which uses this a lot.

I knew exactly where you had gotten it, and I have also filed an issue (#340) against all the others, too.

t4k commented:
That last fix addressed all the issues I found. I'll let @caltechheather poke around on our test instance for another day and then will be ready to say this look good from our end!

caltechheather commented:
@queryluke and @fondrenlibrary we are also seeing the problems with the ISXN that you mentioned in our instance with Caltech. I will do what Luke mentions and report to them. You may be interested to know that @lnarizny and I are meeting with the developers of the RM API in Ipswich next week. I am happy to address all EBSCO-related errors with them on CORAL's behalf.

queryluke commented:
@caltechheather I put in a support ticket but haven't heard anything yet. I'll touch base with Carla and see if she can contact ebsco as well. 

Like I said above, I can put in a band-aid and try to parse the ISxNs with regexp, but I feel it would be better to get the correct data from the API.

jeffnm commented:
When the Ebsco API key was entered into the system and saved into the configuration.ini file, the defaultsort value in our file was written with the escaped quotes unescaped. This may be a result of our specific defaultsort value, but it resulted in 500 errors for the resource module because the database connection failed to load after the syntax error. Did anyone else have this occur?

queryluke commented:
@jeffnm

That's very strange. Do you still have a copy of your original config file that you could post to this thread (any secrets removed, of course)?

Another option would be creating a new table to hold the ebsco information, but I didn't feel like this fit the current workflow for secrets. But I'm open to change it editing the config.ini presents a problem.

t4k commented:
@jeffnm We don't have a defaultsort value in our file, and I don't know what it does, so I wouldn't know how to test that for us.

jeffnm commented:
@queryluke Sure. The defaultsort value probably shouldn't be set in the configuration file anyway, but here's what we have there:

`"\"TRIM(LEADING 'THE ' FROM (TRIM(LEADING 'EL ' FROM (TRIM(LEADING 'L\\\'' FROM (TRIM(LEADING 'LA ' FROM (TRIM(LEADING 'LE ' FROM (TRIM(LEADING 'LES ' FROM (TRIM(LEADING 'DER ' FROM (TRIM(LEADING 'DIE ' FROM (TRIM(LEADING 'DAS ' FROM UPPER(R.titleText)))))))))))))))))) asc\""`

jeffnm commented:
We've got another issue too, when using the Vendor lookup we're getting 
`PHP Parse error:  syntax error, unexpected '.', expecting ',' or ';' in /coral/resources/admin/classes/domain/EbscoKbVendor.php on line 5`

The server we're testing on has PHP 5.5.14

If I remove the `__DIR__.\` from `EbscoKbVendor.php` line 5, we get results for ebsco when searching the Vendors. 

We're still investigating, but we seem to have issues with other searches as well, but they aren't generating PHP errors (at least not at our current log levels). They just aren't finding results or as many results as expected based on your test screenshots.

fondrenlibrary commented:
@jeffnm,  the same errors are come across when searching by Vendor.

queryluke commented:
@jeffnm @fondrenlibrary @t4k 
Really sorry about these bugs that keep cropping up. I'm away this weekend and don't have my dev machine. But I'll address them when I get back Monday evening (and resolve the merge conflict that has arisen as well).

Thanks for being patient and so thorough with this PR!

queryluke commented:
@jeffnm @fondrenlibrary 
The issue with `defaultsort` should now be resolved. Please test.

Also, I couldn't replicate the `EbscoKbVendor.php line 5` issue. I thought the cause might be because I was creating a static value with interpolation. Please give it a try and let me know what happens.

jeffnm commented:
@queryluke your `defaultsort` fix escaped all single and double quotes within the value, and I think it's a working solution. 

Unfortunately, I can't be sure if it's working or not because some of the upstream development branch changes that have been merged in recently don't seem to have 2.1 upgrade code, resulting in missing database tables in my database, which are breaking core Resource Module functionality. 

I'm pretty confident it's the code related to the ResourceAcquisition table, which I think was first introduced https://github.com/coral-erm/coral/commit/801c751a3db3a951a3a08e596add77300561bd21#diff-ff7cebafbff0c74330b18b40bb00d094

I'll open a new issue about that.

jeffnm commented:
@queryluke @veggiematts #350 and #351 cleared my confusion about the database change, but made me realize that your SQL is in the `update_2.1.0.sql` file, which I believe we want to move away from with the 2.1.0 release. Could you look at #351 and change your SQL file to match the `/resources/install/protected/2.1.0/sequence-PRnumber` format?

This should help us as developers work to make sure they have the latest database without running into conflicts as frequently, as well as helping to organize each versions database changes.

jeffnm commented:
@queryluke I want to add that the vendor search seems to have been fixed. All 3 search for options seem to be working for me. I'll see if there is any additional feedback from eresources staff, but from my perspective it's working.

fondrenlibrary commented:
Except for the missing ResourceAcquisition table, I don't see any other problem with this feature.

fondrenlibrary commented:
After manually run the update_Orders.sql , the missing table issue is cleared, though it's irrelevant to this PR.

queryluke commented:
@jeffnm this is changed. I updated the 2.1.0 update scripts to search the entire `install/protected/2.1.0/` directory. I also changed the example in licensing so future versions can use this technique.

However, concerning the `update_Orders.sql` and #350...even when that get's merged it won't work on a fresh installation, because the code in `update_Orders.sql` is not in the base `install.sql`.

This appears to all be related to #302 ? Is anyone working on that? I can probably get a fix in place if you're still accepting PRs.

jeffnm commented:
@queryluke thanks for doing that! You are also reminding me that we talked about using each version's SQL cumulatively, rather than having to edit the `install.sql` for each release.

@jcuenod can you chime in on the question about #302?

jeffnm commented on resources/ajax_forms/getEbscoKbPackageImportForm.php:
```diff
@@ -0,0 +1,358 @@
+    $organization = isset($result['organizationID']) ? (object)['primaryKey' => $result['organizationID'], 'name' => $result['name']] : false;
+
+    $aliasTypeQuery = "SELECT aliasTypeID, shortName
+			FROM coral_organizations.AliasType";
```
This shouldn't be a hard-coded database name. If users have customized their database name during the install process, this causes a fatal error during a package import.

queryluke replied:
ugh, though i'd fixed that. thanks for the catch! It's updated now.

jeffnm commented:
When importing a title, the form asks for an Acquisition Type. However, the imported title does not have an Order once it's imported, so no Acquisition Type is displayed. Should an Order be created during the import?

veggiematts commented:
A default order is created when a resource is submitted in Coral, through the API and through the CSV import tool, so I would say yes.

queryluke commented:
@jeffnm & @veggiematts ok, I'll take a look at the current csv import and follow that logic to create an order.

jeffnm commented:
@queryluke I want to thank you again for all the work you are doing on this! This is a huge enhancement, and a lot of people are going to be very excited by it!

queryluke commented:
@jeffnm you're welcome! I'm really excited that I had the opportunity to help.

Just pushed a fix on the orders. I also updated the package import form. Corrected the option to create a new ORDER for each title (or package only) rather than WORKFLOW for each title (or package only). I must have conflated workflows and orders when I initially developed this.

jcuenod commented:
Okay, sorry I've been so silent. This looks great @queryluke.
One nitpicky issue is that I noticed some weirdness with tabbing (in `resources/install/resources.php`).

To answer the actual question though: I had begun work on #302 but I haven't done anything on it for quite a while. I've attached the new db_tools file (with a .txt extension) which was the key contribution. I didn't know about `glob` which looks useful, wish I'd known about it. But the idea would then be to use the `process_sequenced_sql_files` from `db_tools` and give it a directory. To be honest, I don't know whether this is even tested so that's why this is not a PR and I can't commit to working on it right now so I'm only going to be slowing @queryluke down...

[db_tools.txt](https://github.com/coral-erm/coral/files/1676365/db_tools.txt)

jeffnm commented:
@jcuenod thanks for that info. @queryluke if you are still willing to work on #302, that would be awesome!

queryluke commented:
@jeffnm yeah, I think I can tackle #302. I'm almost finished with it, just have a very weird bug I need to sort out. But otherwise it seems to be working. I should have a PR ready by the end of the week.

queryluke commented:
@jeffnm, just wanted to alert you that my fix for #302 is up as #354.

efrierson commented:
Hey there - 

Trying to install CORAL with this plug in as a fresh install.  Looks like I'm missing the coral_resources.ResourceAcquisition table.  Definitely not in MySQL.  

`PHP Fatal error:  Uncaught exception 'Exception' with message 'There was a problem with the database: Table 'coral_resources.ResourceAcquisition' doesn't exist' in /var/www/html/resources/admin/classes/common/DBService.php:52\nStack trace:\n#0 /var/www/html/resources/admin/classes/common/DBService.php(91): DBService->checkForError()\n#1 /var/www/html/resources/admin/classes/domain/Resource.php(918): DBService->processQuery('SELECT COUNT(DI...', 'assoc')\n#2 /var/www/html/resources/ajax_htmldata/getSearchResources.php(24): Resource->searchCount(Array)\n#3 /var/www/html/resources/ajax_htmldata.php(24): include('/var/www/html/r...')\n#4 {main}\n  thrown in /var/www/html/resources/admin/classes/common/DBService.php on line 52, referer: http://ec2-34-239-111-165.compute-1.amazonaws.com/resources/`

Tried to manually install using the instructions above, but no file by that name existed:
`resources/install/protected/update_2.1.0.sql`

Now, I'm new to CORAL, so I can't clearly tell if this is a problem with CORAL, or this EBSCOKB Integration branch specifically, so feel free to redirect me.

Can confirm I've got mbstring enabled, displayErrors set to off.  I've plugged in an RM API key and I can query the EBSCO KB without a problem.  But I'm getting the same symptoms @caltechheather had; but my errors are pretty clear - missing some MySQL tableage.  Any help our guidance appreciated!

queryluke commented:
@efrierson , you'll need to manually run the `/resources/install/protected/update_Orders.sql` file. 

Also, the problem is more an issue with this branch. The Coral dev team has been discussing the best way to make DB updates between releases (see #302 & #285). I think some changes are going to be applied in 2.1.

However, this branch is meant to be merged into 2.1, so there will likely be a few oddities with it as other 2.1 features are merged. Thus, it's not a true fork or standalone branch. 

Hope that helps!

queryluke commented:
@veggiematts @caltechheather @jeffnm 
In earlier comments you three noted the issue with ISXNs sometimes not being imported. We've heard back from EBSCO and they claim to have fixed the issue with the API. I check a few random titles on my end, and they appear to be working.

techsvcslib commented:
As per @t4k's admonition on coral-admin discussion list, I wanted to comment that after pretty extensive testing of this feature, I am ok with it being merged.

fondrenlibrary commented:
The latest commit was tested successfully yesterday.

veggiematts commented:
I finally tested this!
First of all, good job!

However, I've encountered a few problems during the test:

 - After an install/upgrade, the installer prompts the user to remove write access to the configuration files. But the EBSCO config page in Admin tries to write the values for Customer ID and Api Key to this very config file, and will silently fail will trying to do so. There's something wrong here, that will probably leave the user confused. We should at least display a warning when saving the values fails.

- None of the notes (manual input or additional URLs) were added to the resources. I'll leave a comment in the code to explain why.

- When trying to import a package of 11048 titles, the import failed on the 9th batch (out of 10) (Titles 8961 to 10080) with the following error message: Could not load package titles. I think that my user session has ended during the import, disconnecting me from Coral, and thus preventing the import to complete. That might be a problem.

- And finally, more on a UI/UX point of view, I didn't find a way to go back to a package after viewing its titles. Am I missing something here?

veggiematts commented on resources/ajax_processing/importFromEbscoKb.php:
```diff
@@ -0,0 +1,829 @@
+                $resourceNote->updateDate = date( 'Y-m-d' );
+                $resourceNote->noteTypeID = $noteType->getInitialNoteTypeID();
+                $resourceNote->tabName = $note['tab'];
+                $resourceNote->resourceID = $resource->primaryKey;
```
Since multiple ordering, notes can be tied to a resource or to an order. ResourceNote.resourceID has been replaced with ResourceNote.entityID.
More specifically:
 - entityID should contain Resource.resourceID for the Product tab
 - entityID should contain ResourceAcquisition.resourceAcquisitionID for Acquisitions, Access and Cataloging tabs.

queryluke replied:
@veggiematts I'm still a bit lost on this. I'm using the exact same code that is in the [submitNewResource.php](https://github.com/coral-erm/coral/blob/development/resources/ajax_processing/submitNewResource.php#L70) file.

Is there new code I should be looking at?

queryluke replied:
nvm, I think I sorted it out

queryluke replied:
ok, this is resolved as well

veggiematts commented on resources/ajax_processing/importFromEbscoKb.php:
```diff
@@ -0,0 +1,829 @@
+    $resource->storageLocationID = '';
+    $resource->registeredIPAddresses = '';
+    $resource->providerText	= $providerText;
+    $resource->coverageText = implode('; ', $title->coverageTextArray);
```
coverageText is now located in ResourceAcquisition.

fondrenlibrary commented:
Second @veggiematts 's finding about loading large package. ![image](https://user-images.githubusercontent.com/11461564/36604465-654c094c-1883-11e8-86b8-db66a9c1e995.png)

jeffnm commented:
> After an install/upgrade, the installer prompts the user to remove write access to the configuration files. But the EBSCO config page in Admin tries to write the values for Customer ID and Api Key to this very config file, and will silently fail while trying to do so. There's something wrong here, that will probably leave the user confused. We should at least display a warning when saving the values fails.

I agree that this shouldn't fail silently, but the prompt actually suggests a permission level that is still writable by the owner, which is almost always the web server, so if that chmod command is used this shouldn't normally be an issue for users. Still, failing to write to the config silently is a problem regardless of how the permission issue developed.

queryluke commented:
I'll look into these issues this weekend.

> And finally, more on a UI/UX point of view, I didn't find a way to go back to a package after viewing its titles. Am I missing something here?

I tried to follow the convention of the resource search but am a bit limited by the ajax-y workflow. I could refactor the search into a basic CRUD, and pass parameters via GET method. Or, if everyone would prefer to keep the ajax workflow, I could make 3 separate search pages.

queryluke commented:
@veggiematts @fondrenlibrary @jeffnm @t4k 

Hey all,

This concerns the issue @veggiematts found:
> When trying to import a package of 11048 titles, the import failed on the 9th batch (out of 10) (Titles 8961 to 10080) with the following error message: Could not load package titles. I think that my user session has ended during the import, disconnecting me from Coral, and thus preventing the import to complete. That might be a problem.

I'm hoping to get some advice or ideas. The underlying issue is that the EBSCO API only retrieves a set of 10,000 records. There is an `offset` parameter, but this only offsets within the set of 10,000 records. Whereas, I thought passing something like `offset=501` would pull the 10,001st - 10,020th results. Unfortunately, it does not. In short, if a query has more than 10,000 results then is no possible way to get access to the results beyond 10,000. 

Here are some options, I've come up with, but I'm open to additional ideas or suggestions.

### Further limit the results by resourceType. 
There are 14 different resource types that could split the list into < 10k chunks. But this won't always work. For example, an ebook provider might have over 10k titles with a single 'book' type.

### Ask EBSCO to increase the cap
I'm not sure, but I fear there might be hardware and/or software constraints. It wouldn't hurt to ask, but I've found a package with over 100,000 titles, and I highly doubt they'd increase it that much. Most of the packages I've found tend to be < 20k, so maybe start with that ask, and put a big warning if someone tries to import more than 20k records?

### Ask ebsco for a 'starts with' wildcard search
Right now it's not possible to do a wildcard or starts with search, but, in theory, if they added these we could loop through alpha-numeric searches. This would also require them to develop something and I've no clue how long that would take.

### Attempt to collect all ids
Another approach might be making various queries using 'uncommon' letters to get results under 10k, and make a large collection of title ids in php memory. Once the count of ids in memory matches the total records in the package, then loop through the ids in memory and import the titles. The only problem is that this would drastically increase the execution time.

Please let me know what you think.

cjc66 commented:
Luke, I will forward your thoughts to the EBSCO rep we have chatted with recently to see if he has any suggestions.

-Carla

From: person [mailto:notifications@github.com]
Sent: Monday, February 26, 2018 12:14 PM
To: coral-erm/coral <coral@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [coral-erm/coral] SirsiDynix: EBSCO kb integration (#334)


@veggiematts<https://github.com/veggiematts> @fondrenlibrary<https://github.com/fondrenlibrary> @jeffnm<https://github.com/jeffnm> @t4k<https://github.com/t4k>

Hey all,

This concerns the issue @veggiematts<https://github.com/veggiematts> found:

When trying to import a package of 11048 titles, the import failed on the 9th batch (out of 10) (Titles 8961 to 10080) with the following error message: Could not load package titles. I think that my user session has ended during the import, disconnecting me from Coral, and thus preventing the import to complete. That might be a problem.

I'm hoping to get some advice or ideas. The underlying issue is that the EBSCO API only retrieves a set of 10,000 records. There is an offset parameter, but this only offsets within the set of 10,000 records. Whereas, I thought passing something like offset=501 would pull the 10,001st - 10,020th results. Unfortunately, it does not. In short, if a query has more than 10,000 results then is no possible way to get access to the results beyond 10,000.

Here are some options, I've come up with, but I'm open to additional ideas or suggestions.

Further limit the results by resourceType.

There are 14 different resource types that could split the list into < 10k chunks. But this won't always work. For example, an ebook provider might have over 10k titles with a single 'book' type.

Ask EBSCO to increase the cap

I'm not sure, but I fear there might be hardware and/or software constraints. It wouldn't hurt to ask, but I've found a package with over 100,000 titles, and I highly doubt they'd increase it that much. Most of the packages I've found tend to be < 20k, so maybe start with that ask, and put a big warning if someone tries to import more than 20k records?

Ask ebsco for a 'starts with' wildcard search

Right now it's not possible to do a wildcard or starts with search, but, in theory, if they added these we could loop through alpha-numeric searches. This would also require them to develop something and I've no clue how long that would take.

Attempt to collect all ids

Another approach might be making various queries using 'uncommon' letters to get results under 10k, and make a large collection of title ids in php memory. Once the count of ids in memory matches the total records in the package, then loop through the ids in memory and import the titles. The only problem is that this would drastically increase the execution time.

Please let me know what you think.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/coral-erm/coral/pull/334#issuecomment-368614702>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AQixCP5hNRLJiETANHRpyVaSeMEsedG1ks5tYwIvgaJpZM4Q_PnS>.


[SirsiDynix]<http://www.sirsidynix.com/>


[Facebook]<http://www.facebook.com/SirsiDynix>


[Twitter]<http://twitter.com/sirsidynix>


[LinkedIn]<http://www.linkedin.com/company/sirsidynix>




Carla Clark
PRODUCT MANAGER I
Carla.Clark@sirsidynix.com
800-288-8020 ext. 5734







[Kickstart Your RFP]<https://shop.sirsidynix.com/?src=sdftr>
        GET YOUR POWER OF LIBRARIES GEAR
Shop now for Shirts, Totes, Buttons, Mugs and more!
[Kickstart Your RFP]<https://shop.sirsidynix.com/?src=sdftr>
[Kickstart Your RFP]<http://go.sirsidynix.com/Advocacy-in-the-Public-and-Private-Institutions-SD-Webinar.html?src=sdftr>
        UPCOMING WEBINAR
Advocacy in the Public and Private Institutions | Feb 21st
[Kickstart Your RFP]<http://go.sirsidynix.com/Advocacy-in-the-Public-and-Private-Institutions-SD-Webinar.html?src=sdftr>
[Kickstart Your RFP]<http://go.sirsidynix.com/Connections-Video_Connections-Video-LP-new.html?src=sdftr>
        CHECK OUT OUR VIDEO FOR YOU
Library Connections
[Kickstart Your RFP]<http://go.sirsidynix.com/Connections-Video_Connections-Video-LP-new.html?src=sdftr>


The information contained in this e-mail is confidential, may contain our trade secrets, and may be privileged. It is intended only for the addressee(s) stated above. Any use, dissemination, distribution, publication, or coping of the information contained in this e-mail is strictly prohibited. If you have received this e-mail in error, please immediately notify us by telephone at 801-223-5200 (or by return e-mail) and delete the e-mail from your system. No employee or agent is authorized to conclude any binding agreement on behalf of our Company with another party by email without express written confirmation by an officer of the Company. Thank you for your cooperation.

techsvcslib commented:
My instinct is to plunk for asking for a reasonable increase from EBSCO but you’re right, there are very large packages for which it’s doubtful this will ever work. We could put up warning text. I’m hopeful that more will come to light as a result of Carla’s inquiry.

Steve

On Feb 26, 2018, at 1:23 PM, Carla Clark <notifications@github.com<mailto:notifications@github.com>> wrote:

Luke, I will forward your thoughts to the EBSCO rep we have chatted with recently to see if he has any suggestions.

-Carla

From: person [mailto:notifications@github.com]
Sent: Monday, February 26, 2018 12:14 PM
To: coral-erm/coral <coral@noreply.github.com<mailto:coral@noreply.github.com>>
Cc: Subscribed <subscribed@noreply.github.com<mailto:subscribed@noreply.github.com>>
Subject: Re: [coral-erm/coral] SirsiDynix: EBSCO kb integration (#334)


@veggiematts<https://github.com/veggiematts> @fondrenlibrary<https://github.com/fondrenlibrary> @jeffnm<https://github.com/jeffnm> @t4k<https://github.com/t4k>

Hey all,

This concerns the issue @veggiematts<https://github.com/veggiematts> found:

When trying to import a package of 11048 titles, the import failed on the 9th batch (out of 10) (Titles 8961 to 10080) with the following error message: Could not load package titles. I think that my user session has ended during the import, disconnecting me from Coral, and thus preventing the import to complete. That might be a problem.

I'm hoping to get some advice or ideas. The underlying issue is that the EBSCO API only retrieves a set of 10,000 records. There is an offset parameter, but this only offsets within the set of 10,000 records. Whereas, I thought passing something like offset=501 would pull the 10,001st - 10,020th results. Unfortunately, it does not. In short, if a query has more than 10,000 results then is no possible way to get access to the results beyond 10,000.

Here are some options, I've come up with, but I'm open to additional ideas or suggestions.

Further limit the results by resourceType.

There are 14 different resource types that could split the list into < 10k chunks. But this won't always work. For example, an ebook provider might have over 10k titles with a single 'book' type.

Ask EBSCO to increase the cap

I'm not sure, but I fear there might be hardware and/or software constraints. It wouldn't hurt to ask, but I've found a package with over 100,000 titles, and I highly doubt they'd increase it that much. Most of the packages I've found tend to be < 20k, so maybe start with that ask, and put a big warning if someone tries to import more than 20k records?

Ask ebsco for a 'starts with' wildcard search

Right now it's not possible to do a wildcard or starts with search, but, in theory, if they added these we could loop through alpha-numeric searches. This would also require them to develop something and I've no clue how long that would take.

Attempt to collect all ids

Another approach might be making various queries using 'uncommon' letters to get results under 10k, and make a large collection of title ids in php memory. Once the count of ids in memory matches the total records in the package, then loop through the ids in memory and import the titles. The only problem is that this would drastically increase the execution time.

Please let me know what you think.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/coral-erm/coral/pull/334#issuecomment-368614702>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AQixCP5hNRLJiETANHRpyVaSeMEsedG1ks5tYwIvgaJpZM4Q_PnS>.


[SirsiDynix]<http://www.sirsidynix.com/>


[Facebook]<http://www.facebook.com/SirsiDynix>


[Twitter]<http://twitter.com/sirsidynix>


[LinkedIn]<http://www.linkedin.com/company/sirsidynix>




Carla Clark
PRODUCT MANAGER I
Carla.Clark@sirsidynix.com<mailto:Carla.Clark@sirsidynix.com>
800-288-8020 ext. 5734







[Kickstart Your RFP]<https://shop.sirsidynix.com/?src=sdftr>
GET YOUR POWER OF LIBRARIES GEAR
Shop now for Shirts, Totes, Buttons, Mugs and more!
[Kickstart Your RFP]<https://shop.sirsidynix.com/?src=sdftr>
[Kickstart Your RFP]<http://go.sirsidynix.com/Advocacy-in-the-Public-and-Private-Institutions-SD-Webinar.html?src=sdftr>
UPCOMING WEBINAR
Advocacy in the Public and Private Institutions | Feb 21st
[Kickstart Your RFP]<http://go.sirsidynix.com/Advocacy-in-the-Public-and-Private-Institutions-SD-Webinar.html?src=sdftr>
[Kickstart Your RFP]<http://go.sirsidynix.com/Connections-Video_Connections-Video-LP-new.html?src=sdftr>
CHECK OUT OUR VIDEO FOR YOU
Library Connections
[Kickstart Your RFP]<http://go.sirsidynix.com/Connections-Video_Connections-Video-LP-new.html?src=sdftr>


The information contained in this e-mail is confidential, may contain our trade secrets, and may be privileged. It is intended only for the addressee(s) stated above. Any use, dissemination, distribution, publication, or coping of the information contained in this e-mail is strictly prohibited. If you have received this e-mail in error, please immediately notify us by telephone at 801-223-5200 (or by return e-mail) and delete the e-mail from your system. No employee or agent is authorized to conclude any binding agreement on behalf of our Company with another party by email without express written confirmation by an officer of the Company. Thank you for your cooperation.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/coral-erm/coral/pull/334#issuecomment-368618062>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AC9mt2ygVl8rK3xeVQz6MyleEaYwOahcks5tYwSHgaJpZM4Q_PnS>.

xsong9 commented:
Adding a function to split the file might be another option. For large marc file, we usually use the marc split function in marcedit to split the file into several files. Not sure how complicated it is to implement though. 

Yan

jeffnm commented:
Would everyone be ok merging this PR as it is, perhaps with a warning about the API limit added to the admin page that turns this feature on? 

Then, whatever comes of the conversations with EBSCO, we can make a new PR with any resulting changes?

PaulPoulain commented:
+100


Le 27/02/2018 à 16:19, Jeffrey Mudge a écrit :
>
> Would everyone be ok merging this PR as it is, perhaps with a warning 
> about the API limit added to the admin page that turns this feature on?
>
> Then, whatever comes of the conversations with EBSCO, we can make a 
> new PR with any resulting changes?
>

-- 
Paul Poulain, Associé-gérant / co-owner
BibLibre, Services en logiciels libres pour les bibliothèques
BibLibre, Open Source software and services for libraries

PaulPoulain commented:
@person,

I agree with carla's comment: the API should return all title if we ask 
it to return all title (eventually with an offset, but no reason to 
limit the offset to 10k)

So the way to go is to ask Ebsco to fix this bug !


Le 26/02/2018 à 20:13, person a écrit :
>
> @veggiematts <https://github.com/veggiematts> @fondrenlibrary 
> <https://github.com/fondrenlibrary> @jeffnm 
> <https://github.com/jeffnm> @t4k <https://github.com/t4k>
>
> Hey all,
>
> This concerns the issue @veggiematts <https://github.com/veggiematts> 
> found:
>
>     When trying to import a package of 11048 titles, the import failed
>     on the 9th batch (out of 10) (Titles 8961 to 10080) with the
>     following error message: Could not load package titles. I think
>     that my user session has ended during the import, disconnecting me
>     from Coral, and thus preventing the import to complete. That might
>     be a problem.
>
> I'm hoping to get some advice or ideas. The underlying issue is that 
> the EBSCO API only retrieves a set of 10,000 records. There is an 
> |offset| parameter, but this only offsets within the set of 10,000 
> records. Whereas, I thought passing something like |offset=501| would 
> pull the 10,001st - 10,020th results. Unfortunately, it does not. In 
> short, if a query has more than 10,000 results then is no possible way 
> to get access to the results beyond 10,000.
>
> Here are some options, I've come up with, but I'm open to additional 
> ideas or suggestions.
>
>
>       Further limit the results by resourceType.
>
> There are 14 different resource types that could split the list into < 
> 10k chunks. But this won't always work. For example, an ebook provider 
> might have over 10k titles with a single 'book' type.
>
>
>       Ask EBSCO to increase the cap
>
> I'm not sure, but I fear there might be hardware and/or software 
> constraints. It wouldn't hurt to ask, but I've found a package with 
> over 100,000 titles, and I highly doubt they'd increase it that much. 
> Most of the packages I've found tend to be < 20k, so maybe start with 
> that ask, and put a big warning if someone tries to import more than 
> 20k records?
>
>
>       Ask ebsco for a 'starts with' wildcard search
>
> Right now it's not possible to do a wildcard or starts with search, 
> but, in theory, if they added these we could loop through 
> alpha-numeric searches. This would also require them to develop 
> something and I've no clue how long that would take.
>
>
>       Attempt to collect all ids
>
> Another approach might be making various queries using 'uncommon' 
> letters to get results under 10k, and make a large collection of title 
> ids in php memory. Once the count of ids in memory matches the total 
> records in the package, then loop through the ids in memory and import 
> the titles. The only problem is that this would drastically increase 
> the execution time.
>
> Please let me know what you think.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/coral-erm/coral/pull/334#issuecomment-368614702>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AEeoz94X_unO9JMaFAVf-Q1vDu3fXFhlks5tYwIugaJpZM4Q_PnS>.
>

-- 
Paul Poulain, Associé-gérant / co-owner
BibLibre, Services en logiciels libres pour les bibliothèques
BibLibre, Open Source software and services for libraries

fondrenlibrary commented:
 Merge it.

Mang
On 2/27/2018 9:19 AM, Jeffrey Mudge wrote:
>
> Would everyone be ok merging this PR as it is, perhaps with a warning 
> about the API limit added to the admin page that turns this feature on?
>
> Then, whatever comes of the conversations with EBSCO, we can make a 
> new PR with any resulting changes?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/coral-erm/coral/pull/334#issuecomment-368913057>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AK7jvIZHgcRUCSNRg0KLwnru7sjbj-58ks5tZBz5gaJpZM4Q_PnS>.
>

veggiematts commented:
I'm also ok with merging this as it is, because I also think that the 10k+ problem is on Ebsco's side, and there nothing we can do to fix it nicely on Coral's side.

techsvcslib commented:
I, too, am ok with merging this with the warning info. added to setup for EBSCO KB integration.

queryluke commented:
From what Carla and I heard back from EBSCO, it appears they're not going to make any changes.

The only 'solution' they offered was using the holdings endpoint. But this endpoint has 2 major problems:

1. It only returns selected titles (thus no way to import 'non-selected' titles). 
2. This endpoint returns ALL titles with no way to limit by package (which is why I didn't use it in the first place).

I could 'fake' limiting by a package by looping through the return records and only importing titles with a matching package ID. But I fear this will really slow down the import process. 

I'll try to keep pointing out that this seems like a flaw in the API, but wanted to give everyone an update.

jeffnm commented:
Alright. Thanks everyone for the thorough testing, and most of all thanks to @queryluke and SirsiDynix for the development. I'm merging now. Additional bug fixes and improvements can be processed as new pull requests.
